File: go/user_key_verifier.go
package octoauth

import (
	"context"
	"sort"
)

// apiKeyEndpoint is the octo-server route for user-api-key verification.
const apiKeyEndpoint = "/v1/auth/verify-api-key"

// apiKeyResponse mirrors the JSON body returned by
// POST /v1/auth/verify-api-key(?include=context).
type apiKeyResponse struct {
	UID             string              `json:"uid"`
	SpaceID         string              `json:"space_id"`
	ContextIncluded *bool               `json:"context_included"`
	OwnedBots       map[string][]string `json:"owned_bots"`
}

// userKeyVerifier resolves user API keys (uk_-prefixed, length >= 35). It
// short-circuits obviously malformed keys locally to shield octo-server from
// wave-attack traffic.
//
// The zero value is not usable; construct via [NewUserKeyVerifier].
type userKeyVerifier struct {
	cfg *Config
}

// NewUserKeyVerifier builds a [Verifier] for user API keys. cfg is patched in
// place with defaults and MUST NOT be nil.
func NewUserKeyVerifier(cfg *Config) Verifier {
	applyDefaults(cfg)
	return &userKeyVerifier{cfg: cfg}
}

// Kind implements [Verifier].
func (v *userKeyVerifier) Kind() PrincipalKind { return KindAPIKey }

// Verify implements [Verifier]. It short-circuits empty and too-short keys,
// checks the shared cache under the "k:" prefix, and calls
// POST /v1/auth/verify-api-key with an optional ?include=context query.
func (v *userKeyVerifier) Verify(ctx context.Context, credential string) (*Principal, error) {
	if credential == "" {
		return nil, &Error{
			Kind:     ErrKindInvalidCredential,
			Message:  "empty credential",
			Verifier: KindAPIKey,
		}
	}
	if err := ValidateAPIKey(credential); err != nil {
		return nil, err
	}
	vc := newVerifyContext(v.cfg, KindAPIKey, "k:", credential, v.cfg.APIKeyTTL)
	if p, err, hit := vc.lookupCache(ctx); hit {
		return vc.recordCacheHit(p, err)
	}
	vc.recordCacheMiss()

	endpoint := apiKeyEndpoint
	includeCtx := v.cfg.RequestIncludeContext != nil && *v.cfg.RequestIncludeContext
	if includeCtx {
		endpoint += "?include=context"
	}
	body, err := doVerifyRequest(ctx, v.cfg, endpoint, map[string]string{"api_key": credential}, KindAPIKey)
	if err != nil {
		return vc.finalize(ctx, nil, err)
	}

	var resp apiKeyResponse
	if decodeErr := decodeVerifyResponse(body, &resp, KindAPIKey); decodeErr != nil {
		return vc.finalize(ctx, nil, decodeErr)
	}
	if resp.UID == "" || resp.SpaceID == "" {
		return vc.finalize(ctx, nil, &Error{
			Kind:     ErrKindInvalidCredential,
			Message:  "verify-api-key response missing uid or space_id",
			Verifier: KindAPIKey,
		})
	}

	p := &Principal{
		Kind:    KindAPIKey,
		UID:     resp.UID,
		SpaceID: resp.SpaceID,
		Context: apiKeyContextFromResponse(resp, includeCtx),
	}
	return vc.finalize(ctx, p, nil)
}

// apiKeyContextFromResponse mirrors [sessionContextFromResponse] for the
// api-key realm. When context is included, Spaces is derived from the
// owned_bots map keys so downstream space-membership checks have a canonical
// list.
func apiKeyContextFromResponse(resp apiKeyResponse, includeRequested bool) PrincipalContext {
	if !includeRequested {
		return PrincipalContext{Kind: ContextNotRequested}
	}
	switch {
	case resp.ContextIncluded == nil:
		return PrincipalContext{Kind: ContextUnknownServer}
	case !*resp.ContextIncluded:
		return PrincipalContext{Kind: ContextNotIncluded}
	default:
		return PrincipalContext{
			Kind:             ContextIncluded,
			Spaces:           spacesFromOwnedBots(resp.OwnedBots),
			OwnedBotsBySpace: resp.OwnedBots,
		}
	}
}

// spacesFromOwnedBots returns the sorted list of map keys. Sorting keeps the
// output deterministic across Go versions and across cache hits.
func spacesFromOwnedBots(owned map[string][]string) []string {
	if len(owned) == 0 {
		return nil
	}
	out := make([]string, 0, len(owned))
	for k := range owned {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: ts/src/user-key-verifier.ts
/**
 * userKeyVerifier — resolves user API keys (uk_-prefixed, length >= 35).
 *
 * Short-circuits obviously malformed keys locally to shield octo-server from
 * wave-attack traffic. include=context is sent by default; the `spaces` field
 * of the PrincipalContext is derived from the sorted owned_bots map keys so
 * the output is deterministic across cache hits.
 */

import { OctoAuthError } from './errors.js'
import type { Config } from './config.js'
import { applyDefaults } from './config.js'
import type { Principal, PrincipalContext, Verifier, VerifyOptions } from './verifier.js'
import { KindAPIKey } from './verifier.js'
import { VerifyContext, doVerifyRequest } from './http-verifier.js'
import { validateAPIKey } from './extract.js'

const API_KEY_ENDPOINT = '/v1/auth/verify-api-key'

interface APIKeyResponse {
  uid?: string
  space_id?: string
  context_included?: boolean
  owned_bots?: Record<string, string[]>
}

/**
 * newUserKeyVerifier builds a Verifier for user API keys. cfg is patched in
 * place with defaults.
 */
export function newUserKeyVerifier(cfg: Config): Verifier {
  const resolved = applyDefaults(cfg)
  return {
    kind: KindAPIKey,
    async verify(credential: string, opts?: VerifyOptions): Promise<Principal> {
      if (credential === '') {
        throw new OctoAuthError('invalid-credential', 'empty credential', {
          verifier: KindAPIKey,
        })
      }
      // Local short-circuit for obvious malformation. Throws OctoAuthError.
      validateAPIKey(credential)
      const vc = new VerifyContext(
        resolved,
        KindAPIKey,
        'k:',
        credential,
        resolved.apiKeyTtlMs,
      )
      const cached = await vc.lookupCache()
      if (cached.hit) {
        return vc.recordCacheHit(cached.principal, cached.error)
      }
      vc.recordCacheMiss()

      const endpoint = resolved.requestIncludeContext
        ? API_KEY_ENDPOINT + '?include=context'
        : API_KEY_ENDPOINT
      try {
        const raw = await doVerifyRequest(
          resolved,
          endpoint,
          { api_key: credential },
          KindAPIKey,
          opts?.signal,
        )
        const resp = raw as APIKeyResponse
        if (!resp || !resp.uid || !resp.space_id) {
          throw new OctoAuthError(
            'invalid-credential',
            'verify-api-key response missing uid or space_id',
            { verifier: KindAPIKey },
          )
        }
        const p: Principal = {
          kind: KindAPIKey,
          uid: resp.uid,
          spaceId: resp.space_id,
          context: apiKeyContextFromResponse(
            resp,
            resolved.requestIncludeContext,
          ),
        }
        return vc.finalizeSuccess(p)
      } catch (err) {
        return vc.finalizeError(err)
      }
    },
  }
}

function apiKeyContextFromResponse(
  resp: APIKeyResponse,
  includeRequested: boolean,
): PrincipalContext {
  if (!includeRequested) return { kind: 'not-requested' }
  if (resp.context_included === undefined) return { kind: 'unknown-server' }
  if (resp.context_included === false) return { kind: 'not-included' }
  const ctx: PrincipalContext = { kind: 'included' }
  if (resp.owned_bots !== undefined) {
    ctx.ownedBotsBySpace = resp.owned_bots
    ctx.spaces = spacesFromOwnedBots(resp.owned_bots)
  }
  return ctx
}

/**
 * spacesFromOwnedBots returns the sorted list of map keys. Sorting keeps the
 * output deterministic across cache hits.
 */
function spacesFromOwnedBots(owned: Record<string, string[]>): string[] {
  const keys = Object.keys(owned)
  keys.sort()
  return keys
}
